File: emisor/capas/presentacion.py
"""Capa de presentacion: texto <-> bits ASCII."""

BITS_POR_CARACTER = 8
LIMITE_ASCII = 128


def codificar_mensaje(texto: str) -> str:
    """Convierte texto a una cadena de bits ASCII (8 bits por caracter)."""
    for caracter in texto:
        if ord(caracter) >= LIMITE_ASCII:
            raise ValueError(f"caracter no ASCII: {caracter!r}")
    return "".join(format(ord(c), f"0{BITS_POR_CARACTER}b") for c in texto)


def decodificar_mensaje(bits: str) -> str:
    """Convierte una cadena de bits ASCII de vuelta a texto."""
    if len(bits) % BITS_POR_CARACTER != 0:
        raise ValueError("la longitud de bits debe ser multiplo de 8")
    caracteres = (bits[i : i + BITS_POR_CARACTER] for i in range(0, len(bits), BITS_POR_CARACTER))
    return "".join(chr(int(byte, 2)) for byte in caracteres)

File: receptor/capas/presentacion.js
/** Capa de presentacion: texto <-> bits ASCII. */

export const BITS_POR_CARACTER = 8;
const LIMITE_ASCII = 128;

/** @param {string} texto @returns {string} */
export function codificarMensaje(texto) {
  for (const caracter of texto) {
    if (caracter.charCodeAt(0) >= LIMITE_ASCII) {
      throw new Error(`caracter no ASCII: ${caracter}`);
    }
  }
  return [...texto].map((c) => c.charCodeAt(0).toString(2).padStart(BITS_POR_CARACTER, "0")).join("");
}

/** @param {string} bits @returns {string} */
export function decodificarMensaje(bits) {
  if (bits.length % BITS_POR_CARACTER !== 0) {
    throw new Error("la longitud de bits debe ser multiplo de 8");
  }
  let texto = "";
  for (let i = 0; i < bits.length; i += BITS_POR_CARACTER) {
    texto += String.fromCharCode(parseInt(bits.slice(i, i + BITS_POR_CARACTER), 2));
  }
  return texto;
}
